### DISCUSSION TITLE
Based on highlighted color , got the header name

### DISCUSSION THREAD

**Discussion by yonglee7015** [pymupdf/PyMuPDF#2156] [Q&A] [answered]
How can I get the header name based on highlighted column. For example,  if Medium column is highlighted ,then get the name of Medium. if Large column is highlighted, the get the name of Large. Can someone give me a hint? 

![image](https://user-images.githubusercontent.com/59268746/210296688-043963b8-9872-4f8a-bc5a-0ad3e76ca6a4.png)

**Comment by JorjMcKie** [comment, marked as answer]
You are showing a PDF viewer (Adobe Acrobat?) dialog, right?
In this environment everything is under that viewer's control, so there no active PyMuPDF.

**Reply by JorjMcKie** [reply]
The world would be different, if you are referring to the PDF being **stored** with a certain column highlighted.
That would open some ways to find out - so please elaborate on this.

**Reply by yonglee7015** [reply]
yes，just as you say it's a pdf wilth certain column highlighted.   so there is any ways to figure it out?  Based on filled color of rects to find the most frequent x0, x1 ?

**Reply by JorjMcKie** [reply]
Ah ok then. This should be possible, but requires some work, because there is no highlevel function for table access.
First thing to do is extracting the vector graphics on the page: `paths = page.get_drawings()`.
From there, select large vertical rectangles in the hope, this will be the backgrounds of your columns. You will need to experiment a little here.
Please make sure to have a look at the docu of `page.get_drawings()`.
```python
paths = page.get_drawings()
vert_rects = []
for path in paths:
    fill = path["fill"]  # path fill color
    for item in path["items"]:  # the drawing primitives in this path
        if item[0] != "re":  # no rectangle, skip
            continue
        rect = item[1]  # the rectangle itself, check its size
        if rect.width < 2 or rect.height < threshold:
            continue  # must be a very high rectangle, but not a vertical line
        vert_rects.append((fill, rect))
```
So we should have a list of rectangles with sufficient height, which also are no rectangles that simulate a line (happens often!).
`threshold` for height could be something like 50% of the page height if your table covers the page ...
Hopefully you will end up with the rects of your table.
Compare the `fill` color and match the table header text coordinates (x0, x1) with the respecive column rectangles ...

**Reply by yonglee7015** [reply]
Thanks,  Cos highlighted column is divided into small rects,  so  rect.height < threshold will ignore all these small rects.

**Reply by JorjMcKie** [reply]
ok, then choose another threshold value, or omit this part of the check

**Reply by yonglee7015** [reply]
Finally I collect all  the  highlighted rects in list (Highlighted rects) , find the most frequency  x0,x1.  Loop in Highlighted rects list and  find the rect which has min y0. Then based on this rec , offset some pixies to locate the header of table.

```
for path in paths:
    fill = path["fill"]  # path fill color
    for item in path['items']:
        if item[0] != 're':
           continue
        rect = item[1]   
        if rect.width <2:   
           continue
        text = page.get_text('text',Rect(rect[0],rect[1]-8,
                       rect[2] + 8,rect[3]))
        try:
            if text and fill > (0,0,0):
               rects.append(rect)
               v_x0_x1.append((rect[0],rect[2]))   
        except:
                continue

for element in rects:
    if (element[0],element[2]) == Counter(v_x0_x1).most_common(1)[0][0]:
        highlight_col.append(element)

min_top = min(map(lambda x: x[1], highlight_col))
sample_rect = [top_rec for top_rec in highlight_col if top_rec[1]== min_top] 
text = page.get_text('text',Rect(sample_rect[0][0],sample_rect[0][1]-30,
                     sample_rect[0][2],sample_rect[0][3]-20))


print(text)
```
